### PR TITLE
Ability to limit to configured application region only

### PR DESCRIPTION
**Pull request by ealsur**
# Pull Request Template

## Description

This PR adds an optional parameter to `CosmosClientBuilder.WithApplicationRegion` that turns off endpoint discovery and makes a multi region account behave as a single region, to enable custom synchronization scenarios.

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -340,6 +340,13 @@ public CosmosSerializer Serializer
             }
         }
 
+        /// <summary>
+        /// When limiting to only the specified region, availability is reduced to that of a single region account.
+        /// </summary>
+        /// <seealso cref="CosmosClientOptions.ApplicationRegion"/>
+        /// <seealso href="https://docs.microsoft.com/azure/cosmos-db/high-availability"/>
+        internal bool LimitToApplicationRegionOnly { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: AllowNonApplicationRegions { get; set; } = true;

Better alternative for  'NonApplication' would be good.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: * AllowApplicationRegionFallback
* AllowApplicationRegionFailover
* AllowApplicationRegionRedundancy

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Fail-over is typically application moved out. 
Fallback: Does fall-back mean a single step fallback? Its still a confusion for Failover

AllowGlobalDistribution
AllowNonApplicationRegions

@kirillg thoughts?

File: changelog.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [#716](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/716) Added camel case serialization on LINQ query generation
 - [#729](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/729) Added aggregate(CountAsync/SumAsync etc.) extensions for LINQ query
 - [#743](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/743) Added WebProxy to CosmosClientOptions
+- [#767](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/767) Added support for custom synchronization of clients
```
Comment (kirankumarkolli) [inline changelog.md]:
```suggestion
- [#767](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/767) Ability to limit to configured application region only
```

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Renamed to UseAnyAccountRegion

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 3]
```diff
@@ -350,6 +350,18 @@ public CosmosSerializer Serializer
 #endif
         bool AllowBulkExecution { get; set; }
 
+        /// <summary>
+        /// Gets or sets the flag to enable endpoint discovery for geo-replicated database accounts in the Azure Cosmos DB service.
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]:
```suggestion
        /// Allow using account region in-case ApplicationRegion is not available. Defaults to 'true'.
```

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 4]
```diff
@@ -350,6 +350,18 @@ public CosmosSerializer Serializer
...
+        /// </summary>
+        /// <value>
+        /// Default value is true.
+        /// </value>
+        /// <remarks>
+        /// When the value of this property is true, the SDK will automatically discover the current write and read regions to ensure requests are sent to the correct region based on the region specified in the ApplicationRegion property. When setting the value to false, availability is reduced to that of a single region account to the region specified in ApplicationRegion.</remarks>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]:
```suggestion
        /// When the value of this property is true, the SDK will automatically discover all account write and read regions, and use them when the configured application region is not available. When set to false, availability is limited to the configured application region only. </remarks>
```

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 5]
```diff
@@ -350,6 +350,18 @@ public CosmosSerializer Serializer
...
+        /// <seealso cref="CosmosClientOptions.ApplicationRegion"/>
+        /// <seealso href="https://docs.microsoft.com/azure/cosmos-db/high-availability"/>
+        internal bool UseAnyAccountRegion { get; set; } = true;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: How about "AllowAnyAccountRegion"?

**Pull request merged**